README.md
# Flask-ApiExceptions

Flask-ApiExceptions is a Flask extension that provides the basic functionality for serializing uncaught exceptions as HTTP responses for a JSON-based REST API.

## Installation

You can install this extension with `pip`:

```bash
$ pip install flask_apiexceptions
```

Or, you can clone the repository:

```bash
$ git clone https://github.com/jperras/flask_apiexceptions.git
```

## Running the Tests

[Tox](https://pypi.python.org/pypi/tox) is used to run the tests, which are
written using [PyTest](https://docs.pytest.org/en/latest/). To run them, clone
the repository (indicated above), ensure `tox` is installed and available, and
run:

```bash
$ cd path/to/flask_apiexceptions
$ tox
```

## Usage

This package includes an extension named `JSONExceptionHandler`, which can be
added to your application in the usual way:

    from flask import Flask
    from flask_apiexceptions import JSONExceptionHandler

    app = Flask(__name__)
    exception_handler = JSONExceptionHandler(app)

The extension can also be initialized via deferred application init if you're
using an application factory:

    exception_handler = JSONExceptionHandler()
    exception_hander.init_app(app)

Once initialized, the extension doesn't actually do anything by default. You'll
have to configure it to handle Werkzeug HTTP error codes or custom `Exception`
classes.

### Custom Exception Class Handling

An example showing how we can raise a custom exception within a view method, and have that exception be transformed into a JSON response:

```python
class MissingUserError(Exception):
    status_code = 404
    message = 'No such user exists.'

@app.route('/not-found')
def testing():
    raise MissingUserError()

ext = JSONExceptionHandler(app)
ext.register(code_or_exception=MissingUserError)

with app.app_context():
    with app.test_client() as c:
        rv = c.get('/not-found')

assert rv.status_code == 404
assert rv.headers['content-type'] == 'application/json'
assert json.loads(rv.data)['message'] == 'No such user exists.'
```


This uses the `JSONExceptionHandler.default_handler()` to transform the
`CustomError` exception class into a suitable response. It attempts to
introspect the exception instance returned for a `message` or `description`
attribute, and also checks to see if there exists a `status_code` attribute.

If any of those fields are found, the default handler will populate the
response data with the given message, and set the response status code. If no
message or status code is present, a default response of `{"message": "An error
occurred!"}` with an `HTTP/1.1 500 Internal Server Error` status code is set.

If you'd like to handle custom exception classes in a different manner, say
because you have more complex data captured within an exception instance, or
the attributes are not conveniently named `message` or `description`, then you
can specify a custom handler for the exception type:

```python
from flask_apiexceptions import JSONExceptionHandler

app = Flask(__name__)
ext = JSONExceptionHandler(app)

class CaffeineError(Exception):
    teapot_code = 418
    special = {'foo': 'bar'}

def caffeine_handler(error):
    response = jsonify(data=error.special)
    response.status_code = error.teapot_code
    return response

@app.route('/testing')
def testing():
    raise CaffeineError()

ext.register(code_or_exception=CaffeineError, handler=caffeine_handler)

with app.app_context():
    with app.test_client() as c:
        rv = c.get('/testing')

assert rv.status_code == 418
assert rv.headers['content-type'] == 'application/json'
assert json.loads(rv.data)['data'] == CaffeineError.special

```

This is also how, incidentally, you could use a response content type other
than `application/json`. Simply construct your own response object isntead of
using `jsonify()` within your handler, as long as it produces a valid response
as a return value.


## Using `ApiException` and `ApiError` objects

`Flask-ApiExceptions` includes a few convenience classes and a handler method
for setting up structured API error responses. They are entirely optional, but
provide some sane defaults that should cover most situatiosn.

An `ApiException` instance wraps one or more `ApiError` instances. In this
sense the `ApiException` is simply the container for the actual error message.
The `ApiError` instance accepts optional `code`, `message`, and `info`
attributes.

The idea is that the `code` should be an identifier for the type of error, for
example `invalid-data` or `does-not-exist`. The `message` field should provide
a more detailed and precise description of the error. The `info` field can be
used for any additional metadata or unstructured information that may be
required.

The `info` field, if utilized, should contain data that is JSON
serializable.

To use these constructs, you need to register the appropriate exception class
as well as an `api_exception_handler` that is provided for just this purpose:

```python
from flask_apiexceptions import (
    JSONExceptionHandler, ApiException, ApiError, api_exception_handler)

app = Flask(__name__)
ext = JSONExceptionHandler(app)
ext.register(code_or_exception=ApiException, handler=api_exception_handler)

@app.route('/custom')
def testing():
    error = ApiError(code='teapot', message='I am a little teapot.')
    raise ApiException(status_code=418, error=error)


with app.app_context():
    with app.test_client() as c:
        rv = c.get('/custom')

        # JSON response looks like...
        # {"errors": [{"code": "teapot", "message": "I am a little teapot."}]}

assert rv.status_code == 418
assert rv.headers['content-type'] == 'application/json'

json_data = json.loads(rv.data)
assert json_data['errors'][0]['message'] == 'I am a little teapot.'
assert json_data['errors'][0]['code'] == 'teapot'
assert json_data['errors'][0]['info'] is None
```

Note that, when using the `ApiException` and `ApiError` classes, the status
code is set on the `ApiException` instance. This makes more sense when you can
set multiple `ApiError` objects to the same `ApiException`:

```python
from flask_apiexceptions import ApiException, ApiError

# ...

@app.route('/testing')
def testing():
    exc = ApiException(status_code=400)
    invalid_address_error = ApiError(code='invalid-data',
                                     message='The address provided is invalid.')
    invalid_phone_error = ApiError(code='invalid-data',
                                   message='The phone number does not exist.',
                                   info={'area_code': '555'})
    exc.add_error(invalid_address_error)
    exc.add_error(invalid_phone_error)

    raise exc

    # JSON response format:
    # {"errors": [
    #     {"code": "invalid-data", "message": "The address provided is invalid."},
    #     {"code": "invalid-data", "message": "The phone number does not exist.", "info": {"area_code": "444"}}
    # ]}
```


CHANGELOG.md
# Change Log

## [1.1.2](https://github.com/jperras/Flask-ApiExceptions/tree/1.1.2) (2018-07-19)

* Adds .pylintrc configuration file and include it in tox run.
* Set explicit `apiexceptions` logger name for easier logging configuration.

[Full Changelog](https://github.com/jperras/Flask-ApiExceptions/compare/1.1.1...1.1.2)

## [1.1.1](https://github.com/jperras/Flask-ApiExceptions/tree/1.1.1) (2018-07-19)

* Update README with new feature descriptions.

[Full Changelog](https://github.com/jperras/Flask-ApiExceptions/compare/1.1.0...1.1.1)

## [1.1.0](https://github.com/jperras/Flask-ApiExceptions/tree/1.1.0) (2018-02-23)

* Allow more streamlined declaration of ApiException subclasses. 

[Full Changelog](https://github.com/jperras/Flask-ApiExceptions/compare/1.0.6...1.1.0)

## [1.0.6](https://github.com/jperras/Flask-ApiExceptions/tree/1.0.6) (2018-02-21)

* Use ReStructuredText instead of Markdown for README.
* Create ApiError object from ApiException args if present.

[Full Changelog](https://github.com/jperras/Flask-ApiExceptions/compare/1.0.5...1.0.6)

## [1.0.5](https://github.com/jperras/Flask-ApiExceptions/tree/1.0.5) (2018-02-21)

* Fix some release-based semantics around version numbers and build products.

[Full Changelog](https://github.com/jperras/Flask-ApiExceptions/compare/1.0.4...1.0.5)

## [1.0.4](https://github.com/jperras/Flask-ApiExceptions/tree/1.0.4) (2018-02-21)

*  Update setup.cfg and gitignore for package building.

[Full Changelog](https://github.com/jperras/Flask-ApiExceptions/compare/1.0.3...1.0.4)

## [1.0.3](https://github.com/jperras/Flask-ApiExceptions/tree/1.0.3) (2018-02-21)

* Synchronize some version number inconsistencies.

[Full Changelog](https://github.com/jperras/Flask-ApiExceptions/compare/1.0.2...1.0.3)

## [1.0.2](https://github.com/jperras/Flask-ApiExceptions/tree/1.0.2) (2018-02-21)

* Removed invalid classifier that was preventing publication to PyPi.

[Full Changelog](https://github.com/jperras/Flask-ApiExceptions/compare/1.0.1...1.0.2)

## [1.0.1](https://github.com/jperras/Flask-ApiExceptions/tree/1.0.1) (2018-02-21)

* Add metadata necessary for PyPi publication.

[Full Changelog](https://github.com/jperras/Flask-ApiExceptions/compare/1.0.0...1.0.1)

## [1.0.0](https://github.com/jperras/Flask-ApiExceptions/tree/1.0.0) (2018-02-21)


test_apiexceptions.py
"""
test_apiexceptions
~~~~~~~~~~~~~~~~~~

Tests for Flask-APIExceptions extension and related functionality.

"""

import pytest
import flask
from flask import json, jsonify
from flask_apiexceptions import (
    JSONExceptionHandler, ApiException, api_exception_handler, ApiError)


#pylint: disable=redefined-outer-name


@pytest.fixture
def app():
    """Flask application fixture."""

    app = flask.Flask(__name__)
    return app


def test_initialize_extension(app):
    """Ensure extension initialization happens as expected."""

    assert 'apiexceptions' not in getattr(app, 'extensions', dict())

    JSONExceptionHandler(app)

    assert 'apiexceptions' in app.extensions
    assert isinstance(app.extensions['apiexceptions'],
                      JSONExceptionHandler)


def test_deferred_init_app(app):
    """Deferred initialization test."""

    assert 'apiexceptions' not in getattr(app, 'extensions', dict())

    ext = JSONExceptionHandler()
    ext.init_app(app)

    assert 'apiexceptions' in app.extensions
    assert isinstance(app.extensions['apiexceptions'],
                      JSONExceptionHandler)


def test_register_exception_type(app):
    """Register an exception type with the default handler."""

    class CustomError(Exception):
        """Exception subclass for testing purposes."""
        message = 'A custom error message'

    @app.route('/testing')
    def testing():  #pylint: disable=locally-disabled,unused-variable
        """Endpoint that simply raises an exception."""
        raise CustomError()

    ext = JSONExceptionHandler(app)
    ext.register(code_or_exception=CustomError)

    with app.app_context():
        with app.test_client() as c:
            rv = c.get('/testing')

    assert rv.status_code == 500  # Default status code.
    assert rv.headers['content-type'] == 'application/json'
    assert json.loads(rv.data)['message'] == 'A custom error message'


def test_register_exception_code_and_handler(app):
    """Handle Flask 404s with JSONExceptionHandler."""

    ext = JSONExceptionHandler(app)
    ext.register(code_or_exception=404,
                 handler=JSONExceptionHandler.handle_404)

    with app.app_context():
        with app.test_client() as c:
            rv = c.get('/notfound')

    assert rv.status_code == 404
    assert rv.headers['content-type'] == 'application/json'
    assert json.loads(rv.data)['message'] == ('The resource at /notfound could'
                                              ' not be found.')


def test_register_exception_class_and_handler(app):
    """Custom exception handling."""

    class CustomError(Exception):
        """Exception subclass with attributes."""
        teapot_code = 418
        special = {'foo': 'bar'}

    def custom_handler(error):
        """Handler for custom exception subclass."""
        response = jsonify(data=error.special)
        response.status_code = error.teapot_code
        return response

    @app.route('/testing')
    def testing():  #pylint: disable=locally-disabled,unused-variable
        """Endpoint to raise our custom exception."""
        raise CustomError()

    ext = JSONExceptionHandler(app)
    ext.register(code_or_exception=CustomError, handler=custom_handler)

    with app.app_context():
        with app.test_client() as c:
            rv = c.get('/testing')

    assert rv.status_code == 418
    assert rv.headers['content-type'] == 'application/json'
    assert json.loads(rv.data)['data'] == CustomError.special


def test_api_exception_handler(app):
    """Use JSONExceptionHandler.api_exception_handler to capture ApiException
    objects that are raised."""

    @app.route('/custom')
    def testing():  #pylint: disable=locally-disabled,unused-variable
        """Endpoint to raise our custom exception."""
        error = ApiError(code='teapot', message='I am a little teapot.')
        raise ApiException(status_code=418, error=error)

    ext = JSONExceptionHandler(app)
    ext.register(code_or_exception=ApiException, handler=api_exception_handler)

    with app.app_context():
        with app.test_client() as c:
            rv = c.get('/custom')

    assert rv.status_code == 418
    assert rv.headers['content-type'] == 'application/json'

    json_data = json.loads(rv.data)
    assert json_data['errors'][0]['message'] == 'I am a little teapot.'
    assert json_data['errors'][0]['code'] == 'teapot'
    assert json_data['errors'][0]['info'] is None


def test_exception_auto_populate_error():
    """If no ApiError object is provided, create one by default in
    ApiException."""

    exc = ApiException(
        status_code=418,
        code='bad_inputs',
        message='Something happened.',
        info={'key': 'value'})

    assert len(exc.errors) == 1
    assert isinstance(exc.errors[0], ApiError)
    assert exc.errors[0].code == 'bad_inputs'
    assert exc.errors[0].message == 'Something happened.'
    assert exc.errors[0].info == {'key': 'value'}

    exc = ApiException(
        status_code=418,
        code='bad_inputs')

    assert len(exc.errors) == 1
    assert isinstance(exc.errors[0], ApiError)
    assert exc.errors[0].code == 'bad_inputs'
    assert exc.errors[0].message is None
    assert exc.errors[0].info is None

    exc = ApiException(
        status_code=418,
        message='Whoopsie! That is no good.')

    assert len(exc.errors) == 1
    assert isinstance(exc.errors[0], ApiError)
    assert exc.errors[0].info is None
    assert exc.errors[0].code is None
    assert exc.errors[0].message == 'Whoopsie! That is no good.'

    exc = ApiException(
        status_code=418,
        info={'key': 'value'})

    assert len(exc.errors) == 1
    assert isinstance(exc.errors[0], ApiError)
    assert exc.errors[0].info == {'key': 'value'}
    assert exc.errors[0].code is None
    assert exc.errors[0].message is None


def test_api_exception_subclass_variations(app):
    """Subclass ApiException with class attribute descriptors."""

    class CustomError(ApiException):
        """Subclass of ApiException with additional attributes."""
        status_code = 418
        message = "A class attribute exception message."
        code = 'class-attribute'
        info = {'foo': 'bar'}

    @app.route('/testing')
    def testing():  #pylint: disable=locally-disabled,unused-variable
        """Endpoint that simply raises our custom exception."""
        raise CustomError()

    ext = JSONExceptionHandler(app)

    # Use the api_exception_handler since CustomError is a subclass of
    # ApiException
    ext.register(code_or_exception=CustomError, handler=api_exception_handler)

    with app.app_context():
        with app.test_client() as c:
            rv = c.get('/testing')

    assert rv.status_code == 418
    assert rv.headers['content-type'] == 'application/json'
    assert json.loads(rv.data)['errors'] == [
        {'code': CustomError.code,
         'message': CustomError.message,
         'info': CustomError.info}]


flask_apiexceptions.py
# -*- coding: utf-8 -*-

"""
flask_apiexceptions
~~~~~~~~~~~~~~~~~~~

Base API exception classes.

:copyright: (c) Fictive Kin.

"""

from __future__ import absolute_import
from __future__ import print_function

__version_info__ = ('1', '1', '2')
__version__ = '.'.join(__version_info__)
__author__ = 'Joel Perras'
__copyright__ = '(c) 2018 Fictive Kin, LLC'
__all__ = ['JSONExceptionHandler', 'ApiException', 'ApiError',
           'api_exception_handler']


import logging

from flask import jsonify, request
from werkzeug.exceptions import default_exceptions, HTTPException

logger = logging.getLogger('apiexceptions')


class JSONExceptionHandler(object):
    """
    A Flask extension that converts default Flask exceptions to their
    application/json content type equivalent.

    ```
    from application.libs import JSONExceptionHandler
    exception_handler = JSONExceptionHandler()
    exception_handler.init_app(app)
    ```

    """

    # If we don't know what HTTP code to assign an exception, by default
    # we assign it a `500`. This also handles uncaught exceptions; e.g.
    # if our application raises any kind of Exception subclass that we don't
    # explicitly have a handler for, then we've probably got an application
    # error somewhere for that particular code path.
    default_status_code = 500

    default_message = 'An error occurred!'

    def __init__(self, app=None):
        """
        Initialize the extension.

        Any default configurations that do not require the application
        instance should be put here.
        """

        if app:
            self.init_app(app)

    def default_handler(self, error=None):
        """Default error handler to register with the application."""

        if error is None:
            message = self.default_message
        else:
            # If the error object contains a `message` attribute, then let's
            # use that as the message for our exception.
            if hasattr(error, 'message'):
                message = error.message
            # Werkzeug default exception types use `description` instead
            # of `message`.
            elif hasattr(error, 'description'):
                message = error.description
            else:
                message = self.default_message

        response = jsonify(message=message)

        # If our error object contains a specific error code, then let's use
        # that. If not, we will use our `default_status_code` that has been
        # defined for this class. This ensures that random exceptions that
        # are thrown by Python or by external libraries that we miss are
        # an application error.
        response.status_code = self.default_status_code

        if hasattr(error, 'status_code'):
            response.status_code = int(error.status_code)
        elif isinstance(error, HTTPException):
            response.status_code = error.code

        if response.status_code >= 500:
            logger.exception(error)
        else:
            logger.debug(error)

        return response

    def init_app(self, app):
        """
        Initialize the extension with any application-level configuration
        requirements.

        This is where we register the Werkzeug `HTTPException` class along
        with all the other default exception codes.
        """

        self.app = app

        # Register the default HTTP codes to be handled by our exception
        # handler.
        for code, _ in default_exceptions.items():
            self.register(code)

        if not hasattr(self.app, 'extensions'):
            self.app.extensions = {}

        self.app.extensions['apiexceptions'] = self

    def register(self, code_or_exception, handler=None):
        """
        Register an exception class *or* numeric code with the default
        exception handler provided by this extension *or* the function provided
        at `handler` in the argument.

        """
        f = handler or self.default_handler
        self.app.register_error_handler(code_or_exception, f=f)

    @staticmethod
    def handle_404(error=None):  #pylint: disable=locally-disabled,unused-argument
        """The default Werkzeug 404 handler does not include a
        message or description, which causes some consistency issues with our
        frontends when receiving a 404."""

        message = 'The resource at {} could not be found.'.format(request.path)
        response = jsonify(message=message)
        response.status_code = 404
        return response


class ApiError(object):  #pylint: disable=locally-disabled,too-few-public-methods
    """
    Contains information related to an API usage error.

    - code: a semantically readable error slug, e.g. `invalid-password`
    - info: information regarding the source of the error.  E.g., if the
      error was caused by invalid submitted data, the info can contain
      a list of fields that contained the bad data.  E.g.,
      ['username', 'other_field']
    - message: a human-readable description of what went wrong.

    All of the above data will be serialized into JSON to be returned to the
    client.

    """
    def __init__(self, code=None, info=None, message=None):
        self.code = code
        self.info = info
        self.message = message

    def serialize(self):
        """
        Construct response dictionary.
        """
        return {'code': self.code,
                'info': self.info,
                'message': self.message}


class ApiException(Exception):
    """
    An exception that may be raised by various API view endpoints.

    Can contain one or more ApiError objects and must include a HTTP status
    code.
    """

    # Default status code if none is set.
    status_code = 500

    message = None

    code = None

    info = None

    #pylint: disable=locally-disabled,too-many-arguments
    def __init__(self, status_code=None, error=None, message=None, info=None,
                 code=None):
        """
        Initialize the ApiException container object.

        If an `error` instance is provided, it will be added as an error
        contained within this wrapper. If any of `message`, `info`, or `code`
        are set, a new error object is created added.
        """

        super(ApiException, self).__init__()

        self._errors = []
        if error is not None:
            self._errors.append(error)

        self.status_code = status_code or self.status_code
        message = message or self.message
        code = code or self.code
        info = info or self.info

        if message or info or code:
            self.add_error(ApiError(message=message, code=code, info=info))

    def add_error(self, error):
        """
        Append an error to the list of errors contained with this
        ApiException instance.
        """
        self._errors.append(error)

    @property
    def errors(self):
        """Getter for errors currently stored on this instance."""
        return self._errors

    def serialize(self):
        """
        Serialize the errors contained within this ApiException object to
        Python types that are easily convertible to JSON (or similar).
        """
        return {'errors': [e.serialize() for e in self.errors]}


def api_exception_handler(api_exception):
    """
    Jsonify and serialize ApiException-compatible objects and assign
    the correct response code.
    """
    response = jsonify(api_exception.serialize())
    response.status_code = api_exception.status_code
    return response
